agimus_controller/agimus_controller/mpc.py

```diff
@@ -0,0 +1,44 @@
+import time
+from .ocp_base import OCPBase
+from .warm_start_base import WarmStartBase
+from .trajectory import TrajectoryBuffer, TrajectoryPoint
+from .mpc_output import MPCOutputData, MPCDebugData
+
+
+class MPC(object):
+    def __init__(self) -> None:
+        self._ocp = None
+        self._warm_start = None
+        self._trajectory_buffer = TrajectoryBuffer()
+        self._mpc_output_data = MPCOutputData()
+        self._mpc_debug_data = MPCDebugData()
+        self._duration_iteration = 0.0
+        self._duration_horizon_update = 0.0
+        self._duration_generate_warm_start = 0.0
+        self._duration_ocp_solve = 0.0
+
+    def setup(self, ocp: OCPBase, warm_start: WarmStartBase) -> None:
+        self._ocp = ocp
+        self._warm_start = warm_start
+
+    def run(
+        self, initial_state: TrajectoryPoint, trajectory_buffer: TrajectoryBuffer
+    ) -> MPCOutputData:
+        assert self._ocp is not None
+        assert self._warm_start is not None
+        timer1 = time.perf_counter()
+        reference_trajectory = trajectory_buffer.get_points(
+            self._ocp.get_horizon_size()
+        )
+        self._ocp.update_horizon_from_reference(reference_trajectory)
+        timer2 = time.perf_counter()
+        x_init, u_init = self._warm_start.generate(reference_trajectory)
+        timer3 = time.perf_counter()
+        x0 = self._ocp.get_x_from_trajectory_point(initial_state)
```

Comment:
i'm not sure i understand well this line: 

since we have the initial state already, why getting the x0 ?

Reply:
intial state is a `TrajectoryPoint` type. Not a `ndarray`.
While the ocp will expect this transformation to be made.
I could simply send the `TrajectoryPoint` to the ocp and let the creator of the ocp create a method similar to `get_x_from_trajectory_point`.
Maybe I restraining the API a bit too much.
WDYT?

Reply:
No your point totally makes sense. however i'd just change the name "initial_state" to a less missleading one, because you'd think that initial_state is simply an array of (q0,qdot0).

  Maybe could be called "initial_point" ?

Reply:
the state of the robot can be application dependent.
Usually when you talk about the state you are referring to all that is measurable from the robot sensor.
So you do have a "robot_state_estimator".
Then if you consider the state of the OCP it can be different.

On that note I am not sure what is the best way. To be honest `x0 = [q0, v0]` is not always true.
and `x0 = initial_state` modulo some formatting is always true...

Initial point is less obvious to me in terms of optimal control vocabulary...

Reply:
Yes ok i get your point. I thought in the trajectory point we put the weights as well. 

Then if it's only the state, it makes sense. But there shouldn't be the pose of the EE then, because it can be computed through FK

Reply:
ok we can actually use the numpy sensor object from `linear_feedback_controller_msgs` if you prefer. So that is crystal clear it comes from sensor data.

Reply:
Usage of LFC types would make the core MPC depend on ROS package, so I am not sure if this is a good idea. I also still do not understand how it plugs together. x0 should be generic and numpy type. And so should be `x_init` and `u_inti`. But from what you described you need to have internal OCP implementation to get `x0`, but warmstart is then generic or suddenly, there is a strong requirement for the warmstart to have matching types with the OCP?

Additionally if `x0` is transformed with OCP class to be later looped back to different function in the same class without any change sounds... Pointless? Couldn't this transformation happen inside OCP class?

IMO if we stick to the plan where `OCP.solve()` gets only numpy arrays (which I second) I think OCP should never handle data conversions, as those should be agnostic to the case